### PR TITLE
SUS-422 - Chat PHP code cleanup

### PR DESCRIPTION
**Pull request by wladekb**
This is an effort to improve the current state of PHP chat code. This is not intended to implement any new feature or fix existing bug.

WORK IN PROGRESS

Lot of renames and refactoring will occur before I'm done. You're still welcome to review the code as I'm going through this process. Let me know if you want to be pinged when this is moreless completed.

https://wikia-inc.atlassian.net/browse/SUS-422
https://wikia-inc.atlassian.net/browse/SUS-360
https://wikia-inc.atlassian.net/browse/SUS-361

Depends on:
- https://github.com/Wikia/chat/pull/25

/cc @dianafa @mixth-sense @owend @macbre

### CODE REVIEW

File: extensions/wikia/Chat2/ChatServerApiClient.class.php [hunk 1]
```diff
@@ -8,24 +9,27 @@
...
 	 * that the connection is attempting to be made from (prevents
 	 * circumventing bans by connecting to Wiki A's chat via Wiki B).
 	 */
-	static public function getCityIdForRoom( $roomId ) {
+	static public function getCityIdFromRoomId( $roomId ) {
 		wfProfileIn( __METHOD__ );
 
 		$cityId = "";
-		$cityJson = NodeApiClient::makeRequest( array(
+		$cityData = ChatServerApiClient::makeRequest( array(
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatServerApiClient.class.php]: Should use short array notation here?

File: extensions/wikia/Chat2/ChatServerApiClient.class.php [hunk 2]
```diff
@@ -41,45 +45,59 @@ static public function getCityIdForRoom( $roomId ) {
...
-				"roomUsers" => json_encode( $roomUsers ),
-				"extraDataString" => $extraDataString
-			) );
-
-			$roomData = json_decode( $roomJson );
-		}
+		$roomId = null;
+
+		// Add some extra data that the server will want in order to store it in the room's hash.
+		$extraData = array(
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatServerApiClient.class.php]: Short notation

File: extensions/wikia/Chat2/ChatServerApiClient.class.php [hunk 3]
```diff
@@ -41,45 +45,59 @@ static public function getCityIdForRoom( $roomId ) {
...
+			'wgServer' => $wgServer,
+			'wgArticlePath' => $wgArticlePath
+		);
+		$extraDataString = json_encode( $extraData );
+
+		$roomData = ChatServerApiClient::makeRequest( array(
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatServerApiClient.class.php]: Short notation again

File: extensions/wikia/Chat2/ChatServerApiClient.class.php [hunk 4]
```diff
@@ -135,14 +154,19 @@ static private function makeRequest( $params ) {
 		// operation, abort trying to contact the node server since it could be unavailable (in the event of complete
 		// network unavailability in the primary datacenter). - BugzId 11047
 		if ( empty( $wgReadOnly ) ) {
-			$requestUrl = "http://" . NodeApiClient::getHostAndPort() . "/api?" . http_build_query( $params );
+			$requestUrl = "http://" . ChatServerApiClient::getHostAndPort() . "/api?" . http_build_query( $params );
 			$response = Http::get( $requestUrl, 'default', array( 'noProxy' => true ) );
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatServerApiClient.class.php]: What about short notation for the last argument to this `Http::get` call?

File: extensions/wikia/Chat2/Chat.class.php [hunk 5]
```diff
@@ -58,8 +54,9 @@ public static function echoCookies() {
 		if ( !$wg->User->isLoggedIn() ) {
 			return [ "key" => false ];
 		}
-		$key = "Chat::cookies::" . sha1( $wg->User->getId() . "_" . microtime() . '_' .  mt_rand() );
-		$wg->Memc->set( $key, [ "user_id" => $wg->User->getId(), "cookie" => $_COOKIE ] , 60 * 60 * 48 );
+		$key = "Chat::cookies::" . sha1( $wg->User->getId() . "_" . microtime() . '_' . mt_rand() );
+		$wg->Memc->set( $key, [ "user_id" => $wg->User->getId(), "cookie" => $_COOKIE ], 60 * 60 * 48 );
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php]: this looks really messy - what about moving `60 * 60 * 48` and `Chat::cookies::` to consts, maybe some `implode` with `_`  and each param of `set()` can be moved to a separate line.
What would you say about even moving the key creation to a separate function?

File: extensions/wikia/Chat2/Chat.class.php [hunk 6]
```diff
@@ -131,35 +128,18 @@ public static function blockPrivate( $username, $dir = 'add', $kickingUser ) {
...
-							'cbu_user_id' => $kickingUserId,
-							'cbu_blocked_user_id' => $userToBlock->getId()
-						],
-						__METHOD__,
-						[ 'IGNORE' ]
-					);
-				}
-				$dbw->commit();
+		if ( !empty( $blockedUser ) && !$kickingUser->isAnon() ) {
+			$chatUser = new ChatUser( $kickingUser );
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php]: will `ChatUser` be always correctly created?

File: extensions/wikia/Chat2/Chat.class.php [hunk 7]
```diff
@@ -291,150 +224,36 @@ protected static function getTimeLabel( $time ) {
...
-	 * @return bool|object
-	 *
-	 * @throws MWException
-	 */
-	public static function getBanInformation( $cityID, User $banUser ) {
-		if ( $banUser->isAnon() ) {
-			return false;
+	private static function getUserNamesFromIds( $userIds ) {
+		if ( !is_array( $userIds ) ) {
+			$userIds = [ ];
```
Comment (kerriamber) [inline extensions/wikia/Chat2/Chat.class.php]: Why not just `return [ ];` here?

File: extensions/wikia/Chat2/ChatServerApiClient.class.php (hunk 1 above)
Comment (wladekb) [inline extensions/wikia/Chat2/ChatServerApiClient.class.php]: Fixed.

File: extensions/wikia/Chat2/ChatServerApiClient.class.php (hunk 2 above)
Comment (wladekb) [inline extensions/wikia/Chat2/ChatServerApiClient.class.php]: Fixed.

File: extensions/wikia/Chat2/ChatServerApiClient.class.php (hunk 3 above)
Comment (wladekb) [inline extensions/wikia/Chat2/ChatServerApiClient.class.php]: Fixed.

File: extensions/wikia/Chat2/ChatServerApiClient.class.php (hunk 4 above)
Comment (wladekb) [inline extensions/wikia/Chat2/ChatServerApiClient.class.php]: Fixed.

File: extensions/wikia/Chat2/Chat.class.php (hunk 5 above)
Comment (wladekb) [inline extensions/wikia/Chat2/Chat.class.php]: Agreed. I mainly focused on general code structure so far. Will get down to details tomorrow. :)

File: extensions/wikia/Chat2/Chat.class.php (hunk 6 above)
Comment (wladekb) [inline extensions/wikia/Chat2/Chat.class.php]: Do you have any particular case in your mind here or just a general question?

File: extensions/wikia/Chat2/Chat.class.php (hunk 7 above)
Comment (wladekb) [inline extensions/wikia/Chat2/Chat.class.php]: Good point.

File: extensions/wikia/Chat2/Chat.class.php [hunk 8]
```diff
@@ -80,26 +89,26 @@ public static function echoCookies() {
...
+		) {
 
-			return wfMessage( 'chat-ban-you-need-permission', $kickBanPermission )
+			return wfMessage( 'chat-ban-you-need-permission', self::CHAT_MODERATOR )
 				->inContentLanguage()->text() . "\n";
 		}
 
 		// Make sure we aren't trying to kick/ban someone who shouldn't be kick/banned
-		if ( $userToKickBan->isAllowed( $kickBanPermission ) && !$kickingUser->isAllowed( 'chatstaff' ) && !$kickingUser->isAllowed( 'chatadmin' ) ) {
+		if ( $userToKickBan->isAllowed( self::CHAT_MODERATOR ) && !$kickingUser->isAllowed( self::CHAT_STAFF ) && !$kickingUser->isAllowed( 'chatadmin' ) ) {
```
Comment (kerriamber) [inline extensions/wikia/Chat2/Chat.class.php]: Shouldn't we extract `'chatadmin'` to a const? `self::CHAT_ADMIN` perhaps?

File: extensions/wikia/Chat2/Chat.class.php [hunk 9]
```diff
@@ -778,4 +605,39 @@ private static function getBanTimeFactors() {
...
+
+		return $chatters;
+	}
+
+	public static function setChatters( $chatters ) {
+		global $wgMemc;
+		wfProfileIn( __METHOD__ );
+
+		$memKey = wfMemcKey( "ChatServerApiClient::getChatters" );
+		$wgMemc->set( $memKey, $chatters, 60 * 60 * 24 );
```
Comment (kerriamber) [inline extensions/wikia/Chat2/Chat.class.php]: Should `60 * 60 * 24` be extracted to a constant?

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 10]
```diff
@@ -22,129 +26,143 @@ class ChatAjax {
...
-		);
+		];
 
 		// Figure out the error message to return (i18n is done on this side).
-		if ( $retVal['isLoggedIn'] === false ) {
-			$retVal['errorMsg'] = wfMsg( 'chat-no-login' );
-		} else if ( $retVal['canChat'] === false ) {
-			$retVal['errorMsg'] = wfMsg( 'chat-you-are-banned-text' );
+		if ( $res['isLoggedIn'] === false ) {
+			$res['errorMsg'] = wfMsg( 'chat-no-login' );
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatAjax.class.php]: This should be changed to use `wfMessage` instead.

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 11]
```diff
@@ -22,129 +26,143 @@ class ChatAjax {
...
+		} else if ( $res['canChat'] === false ) {
+			$res['errorMsg'] = wfMsg( 'chat-you-are-banned-text' );
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatAjax.class.php]: This should be changed to use `wfMessage` instead.

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 12]
```diff
@@ -22,129 +26,143 @@ class ChatAjax {
...
+		if ( $res['canChat'] ) {
 			$roomId = $wgRequest->getVal( 'roomId' );
-			$cityIdOfRoom = NodeApiClient::getCityIdForRoom( $roomId );
-			if ( $wgCityId !== $cityIdOfRoom ) {
-				$retVal['canChat'] = false; // don't let the user chat in the room they requested.
-				$retVal['errorMsg'] = wfMsg( 'chat-room-is-not-on-this-wiki' );
+			$cityIdFromRoom = ChatServerApiClient::getCityIdFromRoomId( $roomId );
+			if ( $wgCityId !== $cityIdFromRoom ) {
+				$res['canChat'] = false; // don't let the user chat in the room they requested.
+				$res['errorMsg'] = wfMsg( 'chat-room-is-not-on-this-wiki' );
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatAjax.class.php]: This should be changed to use `wfMessage` instead.

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 13]
```diff
@@ -22,129 +26,143 @@ class ChatAjax {
...
+
+		if ( !self::authenticateServerOrUser() ) {
+			wfProfileOut( __METHOD__ );
+
+			return [ 'id' => false ];
+		}
 
 		$users = json_decode( $wgRequest->getVal( 'users' ) );
-		$roomId = NodeApiClient::getDefaultRoomId( 'private', $users );
+		if ( !$users || !is_array( $users ) || !in_array($wgUser->getName(), $users) ) {
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatAjax.class.php]: the `in_array` call needs proper spacing: `in_array( $wgUser->getName(), $users )`

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 14]
```diff
@@ -153,48 +171,52 @@ static public function getPrivateRoomID() {
...
 
 		$kickingUser = $wgUser;
 
-		$retVal = array();
+		$res = [ ];
 		$userToBan = $wgRequest->getVal( 'userToBan' );
 		$userToBanId = $wgRequest->getVal( 'userToBanId', 0 );
 
 		if ( !empty( $userToBanId ) ) {
 			$userToBan = User::newFromId( $userToBanId );
```
Comment (sactage) [inline extensions/wikia/Chat2/ChatAjax.class.php]: Is there any reason we can't just do `$userToBan = User::newFromId( $userToBanId )->getName();`? I thought they fixed this in current php versions

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 15]
```diff
@@ -153,48 +171,52 @@ static public function getPrivateRoomID() {
...
-			if ( !empty( $userToBanId ) ) {
-				$userToBan = $userToBan->getName();
-			}
+			$userToBan = $userToBan->getName();
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatAjax.class.php]: Is there any reason we can't just do $userToBan = User::newFromId( $userToBanId )->getName();? I thought they fixed this in current php versions.

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 16]
```diff
@@ -153,48 +171,52 @@ static public function getPrivateRoomID() {
...
 		}
 
 		$mode = $wgRequest->getVal( 'mode', 'private' );
 
 		if ( empty( $userToBan ) ) {
-			$retVal["error"] = wfMsg( 'chat-missing-required-parameter', 'usertoBan' );
+			$res["error"] = wfMsg( 'chat-missing-required-parameter', 'usertoBan' );
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatAjax.class.php]: Should use `wfMessage` instead of `wfMsg`

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 17]
```diff
@@ -208,63 +230,93 @@ static public function getListOfBlockedPrivate() {
...
+
+			return [ 'error' => self::ERROR_NOT_AUTHENTICATED ];
+		}
 
-		$promottingUser = $wgUser;
+		$promotingUser = $wgUser;
 
-		$retVal = array();
+		$res = [ ];
 		$PARAM_NAME = "userToPromote";
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatAjax.class.php]: Why do we even need this variable? In `blockOrBanChat` we do a similar check for the `userToBan` param, but it is just a bare string.

Would extracting this to a const make more sense?

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 18]
```diff
@@ -208,63 +230,93 @@ static public function getListOfBlockedPrivate() {
...
 		$userToPromote = $wgRequest->getVal( $PARAM_NAME );
+
 		if ( empty( $userToPromote ) ) {
-			$retVal["error"] = wfMsg( 'chat-missing-required-parameter', $PARAM_NAME );
+			$res["error"] = wfMsg( 'chat-missing-required-parameter', $PARAM_NAME );
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatAjax.class.php]: `wfMsg` -> `wfMessage` please :)

File: extensions/wikia/Chat2/ChatConfig.class.php [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	public static function getSecretToken() {
+		return ChatConfig::getConfigValue( 'ChatCommunicationToken' );
+	}
+
+	private static function getServersList( $type ) {
+		$servers = self::getConfigValue( $type . 'ChatServers' );
+
+		return $servers[self::getServerBasket()];
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatConfig.class.php:53]: Should this be formatted as `return $servers[ self::getServerBasket() ];`?

File: extensions/wikia/Chat2/ChatConfig.class.php [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+	 *
+	 * @param string $key Key
+	 * @return mixed|false
+	 */
+	private static function getConfigValue( $key ) {
+		global $wgWikiaEnvironment;
+
+		if ( empty( self::$configData ) ) {
+			$configDir = getenv( 'WIKIA_CONFIG_ROOT' );
+			$configFile = $configDir . '/ChatConfig.json';
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatConfig.class.php]: Could we just do `$configFile = getenv( 'WIKIA_CONFIG_ROOT' ) . '/ChatConfig.json';` ?

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 21]
```diff
@@ -208,63 +230,93 @@ static public function getListOfBlockedPrivate() {
...
-				'options' => Chat::GetBanOptions(),
-				'isChangeBan' => $isChangeBan,
-				'isoTime' => $isoTime,
-				'fmtTime' => $fmtTime
-			)
-		);
-		$retVal = array();
-		$retVal['template'] = $tmpl->render( "banModal" );
-		$retVal['isChangeBan'] = $isChangeBan;
+		$tmpl = new EasyTemplate( dirname( __FILE__ ) . '/templates/' );
```
Comment (macbre) [inline extensions/wikia/Chat2/ChatAjax.class.php]: `__DIR__`

File: extensions/wikia/Chat2/ChatConfig.class.php [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+			return self::$configData[$key];
+		}
+
+		return false;
+	}
+
+	/**
+	 * $mode - true = operation, false = failover
+	 */
+	public static function changeMode( $mode = true ) {
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatConfig.class.php]: `$mode` parameter is not used within function body

File: extensions/wikia/Chat2/ChatConfig.class.php [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+	const CHAT_PREVIEW_ENV = 'preview';
+	const CHAT_VERIFY_ENV = 'verify';
+	const CHAT_PRODUCTION_ENV = 'prod';
+
+	const SERVER_TYPE_MAIN = 'Main';
+	const SERVER_TYPE_API = 'Api';
+
+	const VAR_SERVER_BASKET = 'wgChatServersBasket';
+	const VAR_OPERATION_MODE = 'wgChatOperationMode';
+	const CENTRAL_WIKI_ID = 177;
```
Comment (macbre) [inline extensions/wikia/Chat2/ChatConfig.class.php]: Let's use `Wikia::COMMUNITY_WIKI_ID` so that we do not copy it all over our code.

File: extensions/wikia/Chat2/Chat_setup.php [hunk 24]
```diff
@@ -10,33 +10,31 @@
...
-$wgAutoloadClasses['ChatEntryPoint'] = "$dir/ChatEntryPoint.class.php";
+$wgAutoloadClasses['ChatWidget'] = "$dir/ChatWidget.class.php";
+$wgAutoloadClasses['ChatUser'] = "$dir/ChatUser.class.php";
+$wgAutoloadClasses['ChatConfig'] = "$dir/ChatConfig.class.php";
+$wgAutoloadClasses['ChatHooks'] = "$dir/ChatHooks.class.php";
 $wgAutoloadClasses['ChatController'] = "$dir/ChatController.class.php";
 $wgAutoloadClasses['ChatRailController'] = "$dir/ChatRailController.class.php";
 $wgAutoloadClasses['SpecialChat'] = "$dir/SpecialChat.class.php";
-$wgAutoloadClasses['NodeApiClient'] = "$dir/NodeApiClient.class.php";
+$wgAutoloadClasses['ChatServerApiClient'] = "$dir/ChatServerApiClient.class.phpss.php";
```
Comment (macbre) [inline extensions/wikia/Chat2/Chat_setup.php]: `phpss`? A tpyo? :)

File: extensions/wikia/Chat2/Chat.class.php [hunk 25]
```diff
@@ -125,41 +134,24 @@ public static function banUser( $userNameToKickBan, $kickingUser, $time, $reason
...
-				}
-				$dbw->commit();
+		if ( !empty( $blockedUser ) && !$kickingUser->isAnon() ) {
+			$chatUser = new ChatUser( $kickingUser );
+			if ( $dir == 'remove' ) {
+				$chatUser->unblockUser( $blockedUser );
+			} else {
+				$chatUser->blockUser( $blockedUser );
 			}
+			wfGetDB( DB_MASTER, [ ], F::app()->wg->ExternalDatawareDB );
```
Comment (kerriamber) [inline extensions/wikia/Chat2/Chat.class.php]: Why is this here?

File: extensions/wikia/Chat2/ChatServerApiClient.class.php [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+	 * circumventing bans by connecting to Wiki A's chat via Wiki B).
+	 *
+	 * @param int $roomId
+	 * @return string
+	 */
+	static public function getCityIdFromRoomId( $roomId ) {
+		wfProfileIn( __METHOD__ );
+
+		$cityId = null;
+		$cityData = ChatServerApiClient::makeRequest( [
```
Comment (macbre) [inline extensions/wikia/Chat2/ChatServerApiClient.class.php]: `s/ChatServerApiClient/self/g` ?

File: extensions/wikia/Chat2/ChatUser.class.php [hunk 27]
```diff
@@ -0,0 +1,251 @@
...
+		$this->wg->Memc->set( $this->getBanInfoCacheKey(), $banInfo, self::BAN_INFO_TTL );
+	}
+
+	/**
+	 * Get ban status from database slave
+	 *
+	 * @return stdClass|false
+	 */
+	private function getBanInfoFromDb() {
+		$db = wfGetDB( DB_SLAVE, [ ], $this->wg->ExternalDatawareDB );
```
Comment (macbre) [inline extensions/wikia/Chat2/ChatUser.class.php]: Does a small `getDB` helper makes sense here?

File: extensions/wikia/Chat2/ChatHooks.class.php [hunk 28]
```diff
@@ -196,7 +199,7 @@ public static function formatLogEntry( $type, $action, $title, $forUI, $params,
 			$time = $params[2];
 		}
 
-		return wfMsg( 'chat-' . $action . '-log-entry', $link, $time, $endon );
+		return wfMsg( 'chat-' . $action . '-log-entry', $link, $time, $endOnText );
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatHooks.class.php]: Should be `wfMessage`

File: extensions/wikia/Chat2/ChatHooks.class.php [hunk 29]
```diff
@@ -0,0 +1,217 @@
...
+		$modules[$pos] = [ 'ChatRail', 'placeholder', null ];
+
+		wfProfileOut( __METHOD__ );
+
+		return true;
+	}
+
+	public static function onStaffLogFormatRow( $slogType, $result, $time, $linker, &$out ) {
+		if ( $slogType == self::CHAT_FAILOVER_EVENT_TYPE ) {
+			$out = wfMsgExt( 'chat-failover-log-entry', [ 'parseinline' ], [ $time, $result->slog_user_name, $result->slog_user_namedst, $result->slog_comment ] );
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatHooks.class.php]: Should use `wfMessage`

File: extensions/wikia/Chat2/ChatHooks.class.php [hunk 30]
```diff
@@ -0,0 +1,217 @@
...
+						'msgKey' => [
+							'chat-contributions-ban-notice',
+							$nt->getText() # Support GENDER in 'sp-contributions-blocked-notice'
+						],
+						'offset' => '' # don't use $wgRequest parameter offset
+					]
+				);
+			} else {
+				if ( $wgUser->isAllowed( Chat::CHAT_MODERATOR ) && !$user->isAllowed( Chat::CHAT_MODERATOR ) ) {
+					$tools[] = "<a class='chat-change-ban' data-user-id='{$id}' href='#'>" . wfMsg( 'chat-ban-contributions-heading' ) . "</a>";
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatHooks.class.php]: Should use `wfMessage`

File: extensions/wikia/Chat2/ChatHooks.class.php [hunk 31]
```diff
@@ -0,0 +1,217 @@
...
+
+	public static function onLogLine( $logType, $logaction, $title, $paramArray, &$comment, &$revert, $logTimestamp ) {
+		global $wgUser;
+
+		if ( strpos( $logaction, 'chatban' ) === 0 ) {
+			$user = User::newFromId( $paramArray[1] );
+			if ( !empty( $user ) ) {
+				$chatUser = new ChatUser( $user );
+				if ( $chatUser->isBanned() && $wgUser->isAllowed( Chat::CHAT_MODERATOR ) ) {
+					$revert = "(" . "<a class='chat-change-ban' data-user-id='{$paramArray[1]}' href='#'>" . wfMsg( 'chat-ban-log-change-ban-link' ) . "</a>" . ")";
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatHooks.class.php]: use `wfMessage`

File: extensions/wikia/Chat2/Chat.class.php [hunk 32]
```diff
@@ -42,24 +55,21 @@ private static function permissionsService() {
...
 
 		if ( !$wg->User->isLoggedIn() ) {
-			return [ "key" => false ];
+			return '';
 		}
-		$key = "Chat::cookies::" . sha1( $wg->User->getId() . "_" . microtime() . '_' .  mt_rand() );
-		$wg->Memc->set( $key, [ "user_id" => $wg->User->getId(), "cookie" => $_COOKIE ] , 60 * 60 * 48 );
+		$key = 'Chat::cookies::' . sha1( $wg->User->getId() . "_" . microtime() . '_' . mt_rand() );
+		$wg->Memc->set( $key, [
+			'user_id' => $wg->User->getId(),
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php:72]: why not `wgMemc`, `wgUser` here? isn't `F::app()->wg` an old notation, a bit slower than mentioned?

File: extensions/wikia/Chat2/Chat.class.php [hunk 33]
```diff
@@ -72,435 +82,207 @@ public static function echoCookies() {
...
-
-		if ( $action == 'remove' ) {
-			$timeLabel = $endOn = null;
-
-			self::deleteUserBanFromDB( $cityId, $userID );
-		} else {
-			if ( Chat::getBanInformation( $cityId, $banUser ) !== false ) {
-				$action = 'change';
+		$subjectChatUser = new ChatUser( $subjectUser );
+		if ( $action == self::ACTION_ADD ) {
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php]: why not `===` ?

File: extensions/wikia/Chat2/Chat.class.php [hunk 34]
```diff
@@ -72,435 +82,207 @@ public static function echoCookies() {
...
-	public static function banUserDB( $cityId, $banUser, $adminUser, $time, $reason, $action = 'add' ) {
-		if ( empty( $banUser ) || empty( $adminUser ) ) {
-			return;
-		}
-
-		if ( wfReadOnly() ) {
-			return;
-		}
+		$cityId = F::app()->wg->CityId;
+		$action = $time != 0 ? self::ACTION_ADD : self::ACTION_REMOVE;
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php]: why like that? could you please add comment above explaining why we want to remove users if time equals 0 and add if not?
also, could you please re-add params descriptions removed here? `
@param int $cityId The ID of the wikia where the user was banned
@param User $banUser The ID of the user who was banned
@param User $adminUser The ID of the user doing the banning if adding or updating
@param int $time The time in seconds to ban the user if adding or updating
@param string $reason Why the user ban status was changed
@param string $action The operation` where applicable?

File: extensions/wikia/Chat2/Chat.class.php [hunk 35]
```diff
@@ -72,435 +82,207 @@ public static function echoCookies() {
...
 	 * @param int $time
 	 * @param string $reason
 	 *
 	 * @return bool|string Returns true on success, returns an error message as a string on failure.
 	 */
-	public static function banUser( $userNameToKickBan, $kickingUser, $time, $reason ) {
-		ChatHelper::info( __METHOD__ . ': Method called', [
-			'userNameToKickBan' => $userNameToKickBan,
-			'kickingUser' => $kickingUser,
+	public static function banUser( $subjectUserName, User $adminUser, $time, $reason ) {
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php:90]: this function has around 60 lines, what about dividing it into separate functions and unit test them? For example:
- condition from line 103 to function `hasBanPermissions`
- action on add (lines 118-126)
- action on remove (unban) (l. 128-129) and so on

File: extensions/wikia/Chat2/Chat.class.php [hunk 36]
```diff
@@ -72,435 +82,207 @@ public static function echoCookies() {
...
-				$banInfo = self::NO_BAN_INFORMATION;
+		if ( !empty( $blockedUser ) && !$blockedUser->isAnon() && !$requestingUser->isAnon() ) {
+			$requestingChatUser = new ChatUser( $requestingUser );
+			if ( $dir == 'add' ) {
+				$requestingChatUser->blockUser( $blockedUser );
+			} elseif ( $dir == 'remove' ) {
+				$requestingChatUser->unblockUser( $blockedUser );
 			}
+			wfGetLBFactory()->commitMasterChanges();
 		}
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php:174]: looks like a perfect candidate for testing!

File: extensions/wikia/Chat2/Chat.class.php [hunk 37]
```diff
@@ -72,435 +82,207 @@ public static function echoCookies() {
...
-					Chat::makeGroupNameListForLog( $oldGroups ),
-					Chat::makeGroupNameListForLog( $newGroups )
-				], 'chatmoderator' );
-			} else {
-				$errorMsg = wfMessage( "chat-err-no-permission-to-add-chatmod", $CHAT_MOD_GROUP )->text();
-			}
+			$errorMsg = wfMessage( "chat-err-no-permission-to-add-chatmod", self::CHAT_MODERATOR )->text();
 		}
 
 		return ( $errorMsg == "" ? true : $errorMsg );
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php:252]: i don't like functions which are returning mixed types, what do you think about returning always `$errorMsg` string and check for emptiness while processing result?

File: extensions/wikia/Chat2/Chat.class.php [hunk 38]
```diff
@@ -72,435 +82,207 @@ public static function echoCookies() {
...
+	 * @param string $subjectUserName
+	 * @param User $adminUser
 	 *
 	 * @return bool true on success, returns an error message as a string on failure.
 	 */
-	public static function promoteChatModerator( $userNameToPromote, $promotingUser ) {
-		ChatHelper::info( __METHOD__ . ': Method called', [
-			'userNameToPromote' => $userNameToPromote,
-			'promotingUser' => $promotingUser
+	public static function promoteModerator( $subjectUserName, $adminUser ) {
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php:228]: also, it looks like nice to test :)

File: extensions/wikia/Chat2/Chat.class.php [hunk 39]
```diff
@@ -639,7 +359,50 @@ public static function logChatWindowOpenedEvent() {
...
+	 * Partially copied from SpecialUserrights.php
+	 *
+	 * @param User $user
+	 * @param User $doer
+	 * @param Array $attr An array with parameters passed to LogPage::addEntry() according
+	 *                    to description there these are parameters passed later to wfMsg.* functions
+	 * @param String $type
+	 * @param String|null $reason comment added to log
+	 */
+	public static function addLogEntry( $user, $doer, $attr, $type = 'banadd', $reason = null ) {
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php:367]: doer sounds for me very not intuitive, what about `performer` ?

File: extensions/wikia/Chat2/Chat.class.php [hunk 40]
```diff
@@ -711,14 +474,59 @@ public static function canChat( User $userObject ) {
...
 		}
 
-		if ( Chat::getBanInformation( F::app()->wg->CityId, $userObject ) !== false ) {
+		$chatUser = new ChatUser( $userObject );
+		if ( $chatUser->isBanned() ) {
 			return false;
 		}
 
 		return $userObject->isAllowed( 'chat' );
 	}
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php:477]: a lot of stuff is going on in this function, can we split it?

File: extensions/wikia/Chat2/Chat.class.php [hunk 41]
```diff
@@ -711,14 +474,59 @@ public static function canChat( User $userObject ) {
...
+	public static function canKick( User $subjectUser, User $adminUser ) {
+		return (
+			// must be a chat moderator
+			$adminUser->isAllowed( self::CHAT_MODERATOR )
+			&& (
+				!$subjectUser->isAllowed( self::CHAT_MODERATOR )
+				// moderators can be kicked only by chat staff/admins
+				|| $adminUser->isAllowedAny( self::CHAT_STAFF, self::CHAT_ADMIN )
+			) );
+	}
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php:495]: what about unit tests for that?

File: extensions/wikia/Chat2/Chat.class.php [hunk 42]
```diff
@@ -727,42 +535,44 @@ public static function getBanOptions() {
...
+				}
+				$number = (int)trim( $explode2[0] );
+				$factorText = trim( $explode2[1] );
+				$factor = 0;
+
+				if ( isset( $FACTORS[$factorText] ) ) {
+					$factor = $FACTORS[$factorText];
+				} elseif ( isset( $FACTORS[$factorText . 's'] ) ) {
+					$factor = $FACTORS[$factorText . 's'];
+				}
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php]: can we move above lines to a separate function?

File: extensions/wikia/Chat2/Chat.class.php [hunk 43]
```diff
@@ -778,4 +588,41 @@ private static function getBanTimeFactors() {
...
+		wfProfileOut( __METHOD__ );
+	}
+
+	public static function info( $message, Array $params = [ ] ) {
+		\Wikia\Logger\WikiaLogger::instance()->info( 'CHAT: ' . $message, $params );
+	}
+
+	public static function debug( $message, Array $params = [ ] ) {
+		\Wikia\Logger\WikiaLogger::instance()->debug( 'CHAT: ' . $message, $params );
+	}
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php:560]: nice and clean ^ :+1:

File: extensions/wikia/Chat2/Chat.class.php [hunk 44]
```diff
@@ -778,4 +588,41 @@ private static function getBanTimeFactors() {
...
+
+		// data is stored in memcache and set by node.js
+		$chatters = $wgMemc->get( $memcKey );
+		if ( !$chatters ) {
+			$chatters = [ ];
+		}
+
+		wfProfileOut( __METHOD__ );
+
+		return $chatters;
```
Comment (dianafa) [inline extensions/wikia/Chat2/Chat.class.php]: what about `return !$chatters ? [ ] : $chatters`

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 45]
```diff
@@ -22,129 +26,143 @@ class ChatAjax {
 	 * If the user is not allowed to chat, an error message is returned (which can be shown to the user).
 	 */
 	static public function getUserInfo() {
```
Comment (dianafa) [inline extensions/wikia/Chat2/ChatAjax.class.php:28]: this function has 90 lines, what about splitting it into couple of smaller pieces?

File: extensions/wikia/Chat2/Chat.class.php (hunk 32 above)
Comment (wladekb) [inline extensions/wikia/Chat2/Chat.class.php:72]: In fact the `$wg` helper is still the approved way so I'd leave it as it is.

**Comment by wladekb** [comment]
Now part of #10243
